### PR TITLE
chore: convert user erros to amplify errors

### PR DESCRIPTION
**Pull request by akshbhu**
#### Description of changes
- Convert Transformer Errors to Amplify Errors.
- Convert schemaNot Found and Errors thrown from overrides/ add api / update 

##### CDK / CloudFormation Parameters Changed
None

#### Description of how you validated changes
- unit tests implemented and updated
- manul test
- cloud-e2e : https://app.circleci.com/pipelines/github/aws-amplify/amplify-category-api/4543/workflows/7faae165-9f81-48b3-98f2-c617e4f80b72

#### Checklist

- [x] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] Any CDK or CloudFormation parameter changes are called out explicitly

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-category-api/src/category-utils/schema-reader.ts [hunk 1]
```diff
@@ -42,7 +43,10 @@ export class SchemaReader {
     } else if (fs.pathExistsSync(schemaDirPath)) {
       this.schemaPath = schemaDirPath;
     } else {
-      throw new Error(`No schema found, your graphql schema should be in either ${schemaFilePath} or ${schemaDirPath}`);
+      throw new AmplifyError('ApiCategorySchemaNotFoundError', {
+        message: 'No schema found',
+        resolution: `your graphql schema should be in either ${schemaFilePath} or ${schemaDirPath}`,
```
Comment (lazpavel) [inline packages/amplify-category-api/src/category-utils/schema-reader.ts:48]:
```suggestion
        resolution: `Your graphql schema should be in either ${schemaFilePath} or ${schemaDirPath}`,
```

File: packages/amplify-category-api/src/commands/api/add-graphql-datasource.ts [hunk 2]
```diff
@@ -119,7 +119,10 @@ export const run = async (context: $TSContext): Promise<void> => {
 
       fs.writeFileSync(rdsSchemaFilePath, graphql.print(rdsGraphQLSchemaDoc), 'utf8');
     } else {
-      throw new Error(`Could not find a schema in either ${graphqlSchemaFilePath} or schema directory at ${schemaDirectoryPath}`);
+      throw new AmplifyError('ApiCategorySchemaNotFoundError', {
+        message: 'No schema found',
+        resolution: `your rds schema should be in either ${graphqlSchemaFilePath} or schema directory ${schemaDirectoryPath}`,
```
[the remark this answers is not in the document]
Comment (lazpavel) [reply packages/amplify-category-api/src/commands/api/add-graphql-datasource.ts:124]: Consider expanding RDS

File: packages/amplify-category-api/src/graphql-transformer/transform-graphql-schema.ts [hunk 3]
```diff
@@ -11,8 +12,12 @@ export const transformGraphQLSchema = async (
...
-  const transformerVersion = await ApiCategoryFacade.getTransformerVersion(context);
-  return transformerVersion === 2
-    ? transformGraphQLSchemaV2(context, options)
-    : transformGraphQLSchemaV1(context, options);
+  try {
+    const transformerVersion = await ApiCategoryFacade.getTransformerVersion(context);
+    return transformerVersion === 2
+      ? transformGraphQLSchemaV2(context, options)
+      : transformGraphQLSchemaV1(context, options);
+  } catch (error) {
```
[the remark this answers is not in the document]
Comment (akshbhu) [reply packages/amplify-category-api/src/graphql-transformer/transform-graphql-schema.ts:20]: the problem here is we are wrapping userErrors asamplify faults in gql Errors. 
Here we typecast such errors to amplify Errors.
I don't want to have  gql transformers depends on amplify errors. In amplify-category-api package I have changed most of those errors as amplifyErrors.

[the remark this answers is not in the document]
Comment (akshbhu) [reply packages/amplify-category-api/src/graphql-transformer/transform-graphql-schema.ts:20]: here is some context: 
https://github.com/aws-amplify/amplify-cli/pull/12058#discussion_r1113428072

**Review by lazpavel — APPROVED** [review]

File: packages/amplify-category-api/src/commands/api/add-graphql-datasource.ts [hunk 4]
```diff
@@ -233,12 +236,12 @@ export const readSchema = (graphqlSchemaFilePath: string): graphql.DocumentNode
...
-    const error = new Error(`Could not parse graphql schema \n${relativePathToInput}\n${err.message}`);
-
-    error.stack = undefined;
-
-    throw error;
+    // can be changed to GraphQlError when the error type "GraphQlError"  gets updated in amplify-cli-core
+    throw new AmplifyError('UserInputError', {
+      message: `Could not parse graphql scehma \n${relativePathToInput}\n`,
+      details: err.message,
+      link: 'https://docs.amplify.aws/cli-legacy/graphql-transformer/relational/',
```
Comment (phani-srikar) [inline packages/amplify-category-api/src/commands/api/add-graphql-datasource.ts:243]: Is this doc relevant here? Looks like we're pointing to `legacy` docs.

File: yarn.lock [hunk 5]
```diff
@@ -73,6 +73,46 @@
     uuid "^8.3.2"
```
Comment (phani-srikar) [inline yarn.lock:73]: Don't think we need to update these dependencies. These are probably remnants of `setup-dev` run?

File: packages/amplify-category-api/src/index.ts [hunk 6]
```diff
@@ -315,26 +317,32 @@ export const transformCategoryStack = async (context: $TSContext, resource: $TSO
...
-      ]);
-    }
-  } else if (resource.service === AmplifySupportedService.APIGW) {
-    if (canResourceBeTransformed(resource.resourceName)) {
-      // Rebuild CFN
-      const apigwStack = new ApigwStackTransform(context, resource.resourceName);
-      apigwStack.transform();
+        ]);
+      } catch (error) {
+        throw new AmplifyErrorConverter().create(error);
```
Comment (sobolk) [inline packages/amplify-category-api/src/index.ts:338]: Is AmplifyErrorConverter stateless ?
If so I'd rather 
```
import * as AmplifyErrorConverter from './errors/amplify-error-converter';

AmplifyErrorConverter.convertToAmplifyError(error)
```

File: packages/amplify-category-api/src/errors/amplify-error-converter.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import {
+  $TSAny, AmplifyError, AmplifyErrorType,
+} from 'amplify-cli-core';
+
+const amplifyErrorList = [
```
Comment (sobolk) [inline packages/amplify-category-api/src/errors/amplify-error-converter.ts:5]: nit
```suggestion
const amplifyErrorCodes = [
```

Comment (sobolk) [reply packages/amplify-category-api/src/errors/amplify-error-converter.ts:5]: nit 2, consider https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set. for faster lookup.

File: packages/amplify-category-api/src/__tests__/errors/amplify-error-converter.test.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+const amplifyError = new AmplifyError('AmplifyStudioError', {
+  message: 'mockMessage',
+});
+
+test('returns error if the error isnt mentioned in list', async () => {
+  expect(new AmplifyErrorConverter().create(error)).toBeInstanceOf(Error);
+  expect(new AmplifyErrorConverter().create(error).name).toMatch(errorType);
+});
+
+test('returns adefault error if the error isnt instance of Error', async () => {
```
Comment (phani-srikar) [inline packages/amplify-category-api/src/__tests__/errors/amplify-error-converter.test.ts:21]: nit: space `a default`

File: packages/amplify-category-api/src/__tests__/commands/api/add-graphql-datasource.test.ts [hunk 9]
```diff
@@ -1,22 +1,32 @@
-import { readSchema } from '../../../commands/api/add-graphql-datasource';
-import * as path from 'path';
+import { readSchema } from "../../../commands/api/add-graphql-datasource";
```
Comment (sobolk) [inline packages/amplify-category-api/src/__tests__/commands/api/add-graphql-datasource.test.ts:1]: why did we convert single quotes to double in this file?

File: packages/amplify-category-api/src/index.ts (hunk 6 above)
Comment (sobolk) [reply packages/amplify-category-api/src/index.ts:338]: If you intend to keep AmplifyErrorConverter as is you should rename `create` to `convert` .

**Review by phani-srikar — CHANGES_REQUESTED** [review]

File: packages/amplify-category-api/src/commands/api/add-graphql-datasource.ts (hunk 4 above)
Comment (akshbhu) [reply packages/amplify-category-api/src/commands/api/add-graphql-datasource.ts:243]: I couldn't find RDS docs on our new amplify docs site

File: packages/amplify-category-api/src/errors/amplify-error-converter.ts (hunk 7 above)
Comment (akshbhu) [reply packages/amplify-category-api/src/errors/amplify-error-converter.ts]: yeah that is most appropriate for this: https://www.tech-hour.com/javascript-performance-and-optimization

File: yarn.lock (hunk 5 above)
Comment (akshbhu) [reply yarn.lock:73]: This is in main, I guess we can target a separate PR in main for this

File: packages/amplify-category-api/src/provider-utils/awscloudformation/utils/getAppSyncApiName.ts [hunk 10]
```diff
@@ -9,7 +9,10 @@ export const getAppSyncApiResourceName = async (context: $TSContext): Promise<st
     const resource = apiResource[0];
     apiResourceName = resource.resourceName;
   } else {
-    throw new Error(`${AmplifySupportedService.APPSYNC} API does not exist. To add an api, use "amplify update api".`);
+    throw new AmplifyError('NotImplementedError', {
+      message: `${AmplifySupportedService.APPSYNC} API does not exist`,
+      resolution: 'To add an api, use amplify add api',
```
Comment (phani-srikar) [inline packages/amplify-category-api/src/provider-utils/awscloudformation/utils/getAppSyncApiName.ts:14]: nit: surround the command with double quotes.

File: packages/amplify-category-api/src/provider-utils/awscloudformation/cfn-api-artifact-handler.ts [hunk 11]
```diff
@@ -60,7 +60,10 @@ class CfnApiArtifactHandler implements ApiArtifactHandler {
     const meta = stateManager.getMeta();
     const existingApiName = getAppSyncResourceName(meta);
     if (existingApiName) {
-      throw new Error(`GraphQL API ${existingApiName} already exists in the project. Use 'amplify update api' to make modifications.`);
+      throw new AmplifyError('ResourceAlreadyExistsError', {
+        message: `GraphQL API ${existingApiName} already exists in the project`,
+        resolution: 'Use amplify update api to make modifications',
```
Comment (phani-srikar) [inline packages/amplify-category-api/src/provider-utils/awscloudformation/cfn-api-artifact-handler.ts:65]: nit: double quote the commands

**Review by alharris-at — APPROVED** [review]

**Review by phani-srikar — CHANGES_REQUESTED** [review]
Minor nits for readability

**Review by sobolk — APPROVED** [review]

**Review by lazpavel — APPROVED** [review]

**Review by phani-srikar — APPROVED** [review]

**Pull request merged**